interfaces/repo_test.go

```diff
@@ -848,3 +848,179 @@ func (s *RepositorySuite) TestSecuritySnippetsForSnapFailureWithPermanentSnippet
 	c.Assert(err, ErrorMatches, "cannot compute static snippet for consumer")
 	c.Check(snippets, IsNil)
 }
+
+// Tests for AddSnap and RemoveSnap
+
+type AddRemoveSuite struct {
+	repo *Repository
+}
+
+var _ = Suite(&AddRemoveSuite{})
+
+func (s *AddRemoveSuite) SetUpTest(c *C) {
+	s.repo = NewRepository()
+	err := s.repo.AddInterface(&TestInterface{InterfaceName: "iface"})
+	c.Assert(err, IsNil)
+	err = s.repo.AddInterface(&TestInterface{
+		InterfaceName:        "invalid",
+		SanitizePlugCallback: func(plug *Plug) error { return fmt.Errorf("plug is invalid") },
+		SanitizeSlotCallback: func(slot *Slot) error { return fmt.Errorf("slot is invalid") },
+	})
+	c.Assert(err, IsNil)
+}
+
+const testConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [iface]
+`
+const testDifferentIfaceConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [iface]
+plugs:
+    iface:
+        interface: other-iface
+`
+const testDifferentAttrsConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [iface]
+plugs:
+    iface:
+        attr: extra
+`
+const testUnknownIfaceConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [unknown]
+`
+const testInvalidIfaceConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [invalid]
+`
+const testDifferentLabelConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [iface]
+plugs:
+    iface:
+        label: New Label
+`
+const testBareConsumerYaml = `
+name: consumer
+apps:
+    app:
+`
+
+const testProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+`
+const testDifferentIfaceProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+slots:
+    iface:
+        interface: other-iface
+`
+const testDifferentAttrsProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+slots:
+    iface:
+        attr: extra
+`
+const testUnknownIfaceProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [unknown]
+`
+const testInvalidIfaceProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [invalid]
+`
+const testDifferentLabelProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+slots:
+    iface:
+        label: New Label
+`
+const testBareProducerYaml = `
+name: producer
+apps:
+    app:
+`
+
+func (s *AddRemoveSuite) addSnap(c *C, yaml string) (*snap.Info, error) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(yaml))
+	c.Assert(err, IsNil)
+	return snapInfo, s.repo.AddSnap(snapInfo)
+}
+
+func (s *AddRemoveSuite) TestAddSnapAddsPlugs(c *C) {
+	_, err := s.addSnap(c, testConsumerYaml)
+	c.Assert(err, IsNil)
+	// The plug was added
+	c.Assert(s.repo.Plug("consumer", "iface"), Not(IsNil))
+}
+
+func (s *AddRemoveSuite) TestAddSnapPanicsOnExistingSnaps(c *C) {
+	_, _ = s.addSnap(c, testConsumerYaml)
+	_, err := s.addSnap(c, testConsumerYaml)
+	c.Assert(err, ErrorMatches, "cannot add plug consumer.iface: existing plug in the way")
```

Comment:
Please just have one test that overs all possible errors on a single add. It should have:
* One unsupported plug interface
* One unsupported slot interface
* One plug with sanitization error
* One slot with sanitization error
This will cover all code paths in the logic, and the most complex case of the error string.